eras/mary/impl/testlib/Test/Cardano/Ledger/Mary/Arbitrary.hs

```diff
@@ -89,9 +96,67 @@ instance Crypto c => Arbitrary (PolicyID c) where
         , elements hashOfDigitByteStrings
         ]
 
-genMultiAsset :: Crypto c => Gen Integer -> Gen (MultiAsset c)
-genMultiAsset genAmount =
-  MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary genAmount)
+genMultiAssetTriple :: Crypto c => Gen Int64 -> Gen (PolicyID c, AssetName, Int64)
+genMultiAssetTriple genAmount = (,,) <$> arbitrary <*> arbitrary <*> genAmount
+
+-- | The number 910 is chosen with the following rationale.
+--
+-- When we generate a number of MultiAssets all at once, that number happens to have
+-- an implicit upper limit due to the Cardano.Ledger.Mary.Value.{to,from}-based
+-- compacting operation. This operation is also performed when we serialise to and from CBOR.
+--
+-- Refering to the haddock for 'Cardano.Ledger.Mary.Value.to' we surmise that
+--   1. The offsets for AssetName and PolicyID are stored as Word16 (maxBound = 65535).
+--   2. All offsets (including those for AssetName and PolicyID) are relative to the whole
+--      of the representation (binary blob) rather than the start of their respective regions.
+--   3. If the offsets exceed their maxBounds, they will overflow.
+--   4. So, we need to ensure that at least the last of the offsets (AssetName offsets) do
+--      not exceed 65535.
+--   5. With `n` as the total number of assets, the inequality to be satisfied is thus:
+--           8n -- Word64 asset quantities
+--        +  2n -- Word16 policy id offsets
+--        +  2n -- Word16 asset name offsets
+--        + 28n -- 28-byte policy ids
```

Comment:
Absolutely, thanks Alexey! This carried over from the thinking of `triples`. And I think it is easy assuming that each policy id has only one asset. There are some properties in reality that I had thought about:
1. If policies cannot have empty assets, then
  1. number of policies is always <= number of assets
  2. in the worst case number of policies == number of assets
  3. and, this number can be atmost 910.
1. Policies can have empty assets
  1. number of policies can be > number of assets
  2. the equality stands inapplicable
  3. but we can still consider constraining number of policies as well as assets to be < 910, by using the generator for triples, but because, as you pointed out earlier, it is uniform, we can't use only that.

I also wondered whether there was a better way to approach the implementation of generators for this problem than recursively scaling down, like you suggested earlier, but I could not think of any approach yet. Generating policy ids and assets separately then begets the question of how to randomly distribute those assets across the policies. One could do a first pass by perhaps iteratively generating assets by `chooseInt (1, 5)` for each policy, and keeping the count to not exceed a number, but I'm not sure that will be too restrictive and hide good corner cases.